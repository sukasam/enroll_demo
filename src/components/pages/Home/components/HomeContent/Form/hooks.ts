import showToast, {
    dismissAllErrorToasts
} from "Components/shared/ShowToaster";
import { getCountryConfig } from "Constants/countryConfig";
import { Alpha2 } from "Constants/countryConfig/enums";
import languages from "Constants/languages";
import { validateSponsor } from "Hydra/validateSponsor";
import { getFirstValidFullName, getFullName } from "Services/utils/fullName";
import isRefValid from "Services/utils/refValidation";
import { useEffect, useState } from "react";

export function useDebounce<T>(value: T, delay: number): T {
    const [debouncedValue, setDebouncedValue] = useState<T>(value);

    useEffect(() => {
        const timer = setTimeout(() => {
            setDebouncedValue(value);
        }, delay);

        return () => {
            clearTimeout(timer);
        };
    }, [value, delay]);

    return debouncedValue;
}

type CountryConfigLanguage = {
    name: string;
    code: string;
};

type AvailableLanguage = {
    label: string;
    value: string;
    defaultCountry: Alpha2;
};

export default function useAvailableLanguages(
    selectedCountry: Alpha2 | null
): AvailableLanguage[] {
    const [availableLanguages, setAvailableLanguages] =
        useState<AvailableLanguage[]>(languages);

    useEffect(() => {
        if (!selectedCountry) {
            setAvailableLanguages(languages);
            return;
        }

        const selectedCountryData = getCountryConfig(selectedCountry, true);
        const countryLanguages = selectedCountryData?.languages?.map(
            (language: CountryConfigLanguage) => ({
                label: language.name,
                value: language.code.toLowerCase(),
                defaultCountry: selectedCountry
            })
        );

        if (!countryLanguages) {
            setAvailableLanguages(languages);
            return;
        }

        setAvailableLanguages(countryLanguages);
    }, [selectedCountry]);

    return availableLanguages;
}

export const checkSponsor = async (
    sponsorId: string,
    enrollerId: string,
    setFormSponsorId: (sponsorId: string) => void,
    setFormSponsorFullName: (sponsorFullName: string) => void,
    setIsLoading: (isLoading: boolean) => void
): Promise<string | boolean> => {
    setIsLoading(true);
    const { isValid, unicityID, refResults } = await isRefValid(sponsorId);
    if (enrollerId === "") {
        return "home_enroller";
    }
    if (!isValid) {
        setFormSponsorFullName("");
        setIsLoading(false);
        return "error_invalid_sponsor";
    }

    const response = await validateSponsor({
        enroller: enrollerId,
        sponsor: unicityID
    });

    if (response.error) {
        setFormSponsorFullName("");
        setIsLoading(false);
        return "error_invalid_sponsor";
    }

    setFormSponsorId(sponsorId);
    setFormSponsorFullName(getFullName(refResults?.humanName));
    setIsLoading(false);
    return true;
};

export const checkRefID = async (
    enrollerId: string,
    setUnicityId: (enrollerId: string) => void,
    setFullName: (enrollerFullName: string) => void,
    setEnrollerAllFullName: (
        enrollerAllFullName: { firstName?: string; lastName?: string } | null
    ) => void,
    setIsLoading: (isLoading: boolean) => void,
    translate: (key: string) => string,
    hiddenEnrollerId?: boolean
): Promise<string | boolean> => {
    setIsLoading(true);
    try {
        const response = await isRefValid(enrollerId);
        const { isValid, refResults, unicityID } = response;

        if (!isValid) {
            setFullName("");
            setIsLoading(false);
            return hiddenEnrollerId
                ? "error_invalid_sponsor"
                : "error_invalid_referrer";
        }

        setUnicityId(unicityID);
        setEnrollerAllFullName(refResults?.humanName || null);
        setFullName(
            getFirstValidFullName([
                refResults?.preferredName,
                refResults?.humanName
            ])
        );
        setIsLoading(false);
        return true;
    } catch (error) {
        setFullName("");
        setIsLoading(false);
        dismissAllErrorToasts();
        showToast(translate("500_error_description"), "error");
        return "";
    }
};
