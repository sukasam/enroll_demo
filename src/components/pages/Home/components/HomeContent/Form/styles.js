import { css } from "@emotion/react";

export default css`
    position: relative;
    font-family: Inter, Arial, Helvetica, sans-serif;
    color: #153862;
    border: none !important;

    .label {
        font-size: 14px;
        font-weight: 500;
        padding: 10px 10px;
        color: #153862;
    }
    .enroller-id-wrapper {
        justify-content: space-between;
        display: flex;
        align-items: center;
    }
    .id-text-box {
        margin-bottom: 20px;
    }

    .sponsor-id {
        width: 288px;
        display: block;
        margin-bottom: 20px;
    }

    .sponsor-checkbox {
        display: flex;
        align-items: center;
    }
    .sponsor-checkbox label {
        font-size: 13px;
        font-weight: 500;
        padding: 10px 10px;
        color: #91a2b5 !important;
    }

    label {
        font-size: 14px;
        font-weight: 500;
        padding: 10px 10px;
        color: #153862;
    }
`;
