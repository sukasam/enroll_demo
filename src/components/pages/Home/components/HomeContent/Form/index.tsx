/** @jsxImportSource @emotion/react */
import { DevTool } from "@hookform/devtools";
import TextField from "Components/shared/Form/components/TextField";
import PrimaryButton from "Components/shared/PrimaryButton";
import SelectDropdown from "Components/shared/SelectDropdown";
import Spinner from "Components/shared/Spinner";
import StyledCheckbox from "Components/shared/StyledCheckbox";
import { useTranslate } from "Components/shared/Translate";
import {
    getCountries,
    getCountryConfig,
    getDefaultLanguage,
    useCountryConfig
} from "Constants/countryConfig";
import { Alpha2 } from "Constants/countryConfig/enums.js";
import { useUser } from "Contexts/UserContext";
import { useTranslations } from "Contexts/translation";
import { Customer } from "Hydra/getRefID";
import mixpanelService from "Services/mixpanel/initializeMixPanel";
import MixpanelEvent from "Services/mixpanel/mixpanelEvent";
import countryRedirect from "Services/utils/countryRedirect";
import { getFullName } from "Services/utils/fullName";
import { getCookie } from "cookies-next";
import { useRouter } from "next/router";
import { useCallback, useEffect, useState } from "react";
import { FormProvider, useForm } from "react-hook-form";
import RedirectPopup from "./components/RedirectPopup";
import useAvailableLanguages, {
    checkRefID,
    checkSponsor,
    useDebounce
} from "./hooks";
import styles from "./styles.js";

interface FormValues {
    country: string;
    language: string;
    enrollerId: string | null;
    sponsorId?: string | null;
}
interface Props {
    refId: string;
    referrer: Customer;
}

function Form({ referrer, refId }: Props): React.ReactElement {
    const {
        country,
        language,
        setCountry: setCtxCountry,
        setLanguage: setCtxLanguge,
        fetchTranslations
    } = useTranslations();

    const router = useRouter();

    const {
        setEnrollerId: setCtxEnrollerId,
        setEnrollerFullName,
        setEnrollerAllFullName,
        setSponsorId: setCtxSponsorId,
        setSponsorFullName,
        enrollerId,
        sponsorFullName,
        enrollerFullName,
        isUserContextLoading
    } = useUser();

    const formMethods = useForm<FormValues>({
        defaultValues: {
            country: country || getCookie("country") || "US",
            language: getDefaultLanguage(
                (country || getCookie("country") || "US") as Alpha2
            ),
            enrollerId: refId || enrollerId || ""
        },
        mode: "onBlur",
        reValidateMode: "onBlur"
    });

    const {
        unregister,
        handleSubmit,
        clearErrors,
        setValue,
        watch,
        trigger,
        formState
    } = formMethods;
    const formValues = watch();
    const debouncedEnrollerId = useDebounce(formValues.enrollerId, 500);

    const setFormEnrollerId = (value: string): void =>
        setValue("enrollerId", value);

    const setFormSponsorId = (value: string): void =>
        setValue("sponsorId", value);

    const setFormCountry = (value: string): void => setValue("country", value);

    const setFormLanguage = (value: string): void =>
        setValue("language", value);

    const setEnrollerId = (value: string): void => {
        setCtxEnrollerId(value);
        setFormEnrollerId(value);
    };
    const setSponsorId = (value: string): void => {
        setCtxSponsorId(value);
        setFormSponsorId(value);
    };
    const setCountry = (value: string): void => {
        setCtxCountry(value);
        setFormCountry(value);
    };
    const setLanguage = (value: string): void => {
        setCtxLanguge(value);
        setFormLanguage(value);
    };

    const { isValid } = formState;
    const [isReferred] = useState<boolean>(!!getCookie("isReferred"));

    const [redirectPath, setRedirectPath] = useState<string | null>(null);
    const [showSponsor, setShowSponsor] = useState<boolean>(false);
    const [isLoading, setIsLoading] = useState<boolean>(false);
    const [previousCountry, setPreviousCountry] = useState<Alpha2 | null>(null);
    const translate = useTranslate();

    const countryConfig = useCountryConfig();
    const { marketExceptions } = countryConfig || {};
    const { hiddenEnrollerId } = marketExceptions || {};

    const countriesList = getCountries(false)
        .filter(country => country.isSupported || country.redirect)
        .map(country => ({ label: country.name, value: country.alpha2 }))
        .sort((a, b) => a.label.localeCompare(b.label));

    const languageOptions = useAvailableLanguages(formValues.country as Alpha2);
    const availableCountries = getCountries(true)
        .filter(country => country.isSupported)
        .map(country => country.alpha2);

    const handleRedirect = useCallback((): void => {
        const { path } = countryRedirect(
            formValues.country as Alpha2,
            formValues.language,
            "",
            enrollerId || ""
        );
        setRedirectPath(path);
    }, [formValues.language, formValues.country, previousCountry, setValue]);

    const getCountryLanguage = useCallback(
        (country: Alpha2, language: string): string => {
            const selectedCountryData = getCountryConfig(
                country as Alpha2,
                true
            );
            const defaultLanguage =
                selectedCountryData?.languages?.find(lang => lang?.default)
                    ?.code || "en";
            return (
                selectedCountryData?.languages?.find(
                    lang => lang?.code === language
                )?.code || defaultLanguage
            );
        },
        []
    );

    const trackCountry = (newCountry: string, prevCountry: string): void => {
        if (prevCountry && newCountry !== prevCountry) {
            mixpanelService.registerSuperProperties({
                user_market: newCountry,
                user_market_extended: newCountry
            });
            mixpanelService.trackEvent(MixpanelEvent.USER_MARKET_CHANGED, {
                previous_user_market: prevCountry,
                event_location: "home"
            });
        }
    };
    const trackLanguage = (newLanguage: string, prevLanguage: string): void => {
        if (prevLanguage && newLanguage !== prevLanguage) {
            mixpanelService.trackEvent(MixpanelEvent.USER_LANGUAGE_CHANGED, {
                previous_user_language: prevLanguage,
                event_location: "home",
                user_language: newLanguage
            });
            mixpanelService.registerSuperProperties({
                user_language: newLanguage
            });
        }
    };

    useEffect(() => {
        if (isUserContextLoading) return;
        if (refId) {
            setEnrollerId(refId);
            setEnrollerFullName(getFullName(referrer?.humanName));
            setSponsorId(refId);
            setSponsorFullName(getFullName(referrer?.humanName));
        }
    }, [isUserContextLoading, refId, referrer]);

    useEffect(() => {
        setCountry(formValues.country);
        const countryLanguage = getCountryLanguage(
            formValues.country as Alpha2,
            formValues.language
        );
        trackCountry(formValues.country, previousCountry as string);
        setLanguage(countryLanguage);
        if (!availableCountries.includes(formValues.country as Alpha2)) {
            handleRedirect();
        }
    }, [formValues.country]);

    useEffect(() => {
        if (
            formValues.country &&
            availableCountries.includes(formValues.country as Alpha2)
        ) {
            setPreviousCountry(formValues.country as Alpha2);
        }
    }, [formValues.country, availableCountries, setPreviousCountry]);

    useEffect(() => {
        const correctLanguage = getCountryLanguage(
            formValues.country as Alpha2,
            formValues.language
        );
        if (correctLanguage !== language) {
            trackLanguage(correctLanguage, language);
            setLanguage(correctLanguage);
            fetchTranslations(formValues.country, correctLanguage);
            mixpanelService.registerSuperProperties({
                user_language: correctLanguage
            });
        }
    }, [
        formValues.language,
        formValues.country,
        language,
        getCountryLanguage,
        fetchTranslations,
        setLanguage,
        trackLanguage
    ]);

    useEffect(() => {
        const currentCountry = formValues.country as Alpha2;
        const defaultLanguage = getDefaultLanguage(currentCountry);
        setValue("language", defaultLanguage);
    }, [formValues.country, setValue]);

    useEffect(() => {
        if (debouncedEnrollerId) {
            trigger("enrollerId");
        }
    }, [debouncedEnrollerId, trigger]);

    const onSubmit = (): void => {
        setIsLoading(true);
        mixpanelService.registerSuperProperties({
            enroller_id: formValues.enrollerId || "",
            sponsor_id: formValues.sponsorId || formValues.enrollerId || ""
        });

        mixpanelService.trackEvent(MixpanelEvent.HOME_START_COMPLETED);
        router.push("/register");
    };
    return (
        <div css={styles}>
            <FormProvider {...formMethods}>
                <form onSubmit={handleSubmit(onSubmit)}>
                    <SelectDropdown
                        name="country"
                        label={translate("home_select_country")}
                        options={countriesList}
                        rules={{ required: true }}
                        value={watch("country")}
                        data-testid="home_select_country"
                    />
                    <SelectDropdown
                        name="language"
                        label={translate("home_select_language")}
                        options={languageOptions}
                        rules={{ required: true }}
                        value={watch("language")}
                        data-testid="home_select_language"
                    />
                    <div
                        className="enroller-id-wrapper"
                        data-testid="home_enroller_wrapper"
                    >
                        <TextField
                            className="id-text-box"
                            name="enrollerId"
                            data-testid="home_enroller"
                            label={translate(
                                hiddenEnrollerId
                                    ? "home_sponsor"
                                    : "home_enroller"
                            )}
                            type="text"
                            rules={{
                                required: false,
                                validate: async (
                                    value: string
                                ): Promise<string | boolean> =>
                                    new Promise(resolve => {
                                        if (!value) {
                                            setEnrollerFullName("");
                                            resolve(
                                                translate(
                                                    hiddenEnrollerId
                                                        ? "home_sponsor"
                                                        : "home_enroller"
                                                )
                                            );
                                        }
                                        checkRefID(
                                            value,
                                            setEnrollerId,
                                            setEnrollerFullName,
                                            setEnrollerAllFullName,
                                            setIsLoading,
                                            translate,
                                            hiddenEnrollerId
                                        ).then(result => {
                                            if (result === "") {
                                                resolve(true);
                                            } else {
                                                resolve(translate(result));
                                            }
                                        });
                                    })
                            }}
                            disabled={isReferred && isUserContextLoading}
                        />
                        <div
                            className="enrollerName"
                            data-testid="enroller_name"
                        >
                            {enrollerFullName}
                        </div>
                    </div>
                    {!hiddenEnrollerId && (
                        <>
                            <div className="sponsor-checkbox">
                                <StyledCheckbox
                                    id="useSponsor"
                                    name="useSponsor"
                                    onChange={(): void => {
                                        setShowSponsor(!showSponsor);
                                        clearErrors("sponsorId");
                                        unregister("sponsorId");
                                    }}
                                    disabled={isUserContextLoading}
                                />
                                <label
                                    htmlFor="useSponsor"
                                    data-testid="home_sponsor_different_enroller"
                                >
                                    {translate(
                                        "home_sponsor_different_enroller"
                                    )}
                                </label>
                            </div>
                            {showSponsor && (
                                <div
                                    className="enroller-id-wrapper"
                                    data-testid="home_sponsor_wrapper"
                                >
                                    <TextField
                                        name="sponsorId"
                                        label={translate("home_sponsor")}
                                        className="id-text-box"
                                        type="text"
                                        data-testid="home_sponsor"
                                        rules={{
                                            required: false,
                                            validate: (
                                                value: string
                                            ): Promise<string | boolean> =>
                                                new Promise(resolve => {
                                                    if (!value) {
                                                        setSponsorFullName("");
                                                        resolve(
                                                            translate(
                                                                "home_sponsor"
                                                            )
                                                        );
                                                    }
                                                    checkSponsor(
                                                        value,
                                                        formValues.enrollerId ||
                                                            "",
                                                        setSponsorId,
                                                        setSponsorFullName,
                                                        setIsLoading
                                                    ).then(result => {
                                                        resolve(
                                                            translate(result)
                                                        );
                                                    });
                                                })
                                        }}
                                    />

                                    <div
                                        className="enrollerName"
                                        data-testid="sponsor_name"
                                    >
                                        {sponsorFullName}
                                    </div>
                                </div>
                            )}
                        </>
                    )}
                    <PrimaryButton
                        type="submit"
                        id="start_enrollment"
                        data-testid="home_get_started"
                        disabled={!isValid}
                    >
                        {isLoading ? (
                            <Spinner height="32px" />
                        ) : (
                            translate("home_get_started")
                        )}
                    </PrimaryButton>
                </form>
                {redirectPath && (
                    <RedirectPopup
                        path={redirectPath}
                        selectedCountry={formValues.country as Alpha2}
                        onClose={(): void => {
                            setRedirectPath(null);
                            setCountry(previousCountry as string);
                        }}
                    />
                )}
            </FormProvider>
            <DevTool control={formMethods.control} />
        </div>
    );
}

export default Form;
