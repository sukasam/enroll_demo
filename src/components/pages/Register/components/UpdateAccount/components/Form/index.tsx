/** @jsxImportSource @emotion/react */
import PhoneNumberField from "Components/shared/Form/components/PhoneNumberField";
import TextField from "Components/shared/Form/components/TextField";
import { useTranslate } from "Components/shared/Translate";
import { useUser } from "Contexts/UserContext";
import { useCountryConfig } from "Constants/countryConfig";
import { useTranslations } from "Contexts/translation";
import commonConfig from "Constants/countryConfig/commonConfig";
import mapFieldToComponent from "Components/shared/Form/components/MapFieldToComponent";

import showToast from "Components/shared/ShowToaster";
import updateCustomer from "Hydra/updateCustomer";
import { useEffect, useMemo } from "react";
import { FormProvider, useForm } from "react-hook-form";
import styles from "./styles";
import validationRules from "./validationRules";

interface FormValues {
    firstName: string;
    lastName: string;
    fullName: string;
    fullNameEn: string;
    phoneNumber: string;
    email: string;
}

type UpdateAccountFormProps = {
    setIsLoading: (loading: boolean) => void;
    isLoading: boolean;
};

function UpdateAccountForm({
    setIsLoading,
    isLoading
}: UpdateAccountFormProps): JSX.Element {
    const {
        userData,
        isLoggedIn,
        userToken,
        href,
        setUserData,
        goToNextSection
    } = useUser();
    const translate = useTranslate();
    const { country } = useTranslations();
    const countryConfig = useCountryConfig();
    const { marketExceptions } = countryConfig || {};
    const formatPhoneNumber = (phoneNumber: string): string => {
        const cleanedNumber = phoneNumber.replace(/\D/g, "");

        return `+${cleanedNumber}`;
    };

    const form = useForm<FormValues>({
        mode: "onTouched",
        defaultValues: {
            firstName: "",
            lastName: "",
            fullName: "",
            fullNameEn: "",
            phoneNumber: "",
            email: ""
        }
    });

    const {
        handleSubmit,
        formState: { errors },
        control,
        setValue,
        watch
    } = form;

    useEffect(() => {
        if (isLoggedIn) {
            setValue(
                "phoneNumber",
                formatPhoneNumber(userData?.phoneNumber || "")
            );
            setValue("email", userData?.email || "");
            setValue("firstName", userData?.firstName || "");
            setValue("lastName", userData?.lastName || "");
            setValue("fullName", userData?.firstName || "");
            setValue("fullNameEn", userData?.lastName || "");
        }
    }, [isLoggedIn, userData, setValue]);

    const phoneNumber = watch("phoneNumber");

    const shouldUpdate = useMemo(
        () => phoneNumber !== formatPhoneNumber(userData?.phoneNumber || ""),
        [phoneNumber, userData, formatPhoneNumber]
    );

    const updateCustomerFromForm = async (data: FormValues): Promise<void> => {
        try {
            if (!shouldUpdate) {
                goToNextSection();
                return;
            }

            setIsLoading(true);
            const userResponse = await updateCustomer(
                { mobilePhone: data.phoneNumber },
                href || "",
                userToken || ""
            );
            if (userResponse && userData) {
                setUserData({
                    ...userData,
                    phoneNumber:
                        userResponse.mobilePhone || userData.phoneNumber
                });
            } else {
                throw new Error("Failed to update user data");
            }
            showToast(
                translate("notification_success_account_updated"),
                "success"
            );
            setIsLoading(false);
            goToNextSection();
        } catch (error: unknown) {
            const errorMessage =
                error instanceof Error
                    ? error.message
                    : "An unknown error occurred";
            showToast(translate(errorMessage), "error");
            setIsLoading(false);
        }
    };

    const onSubmit = (data: FormValues): void => {
        updateCustomerFromForm(data);
    };

    // Helper: group fields into rows of 2 columns, left-aligned if odd
    const formFields = Array.isArray(marketExceptions?.formUpdateAccount)
        ? marketExceptions.formUpdateAccount
        : commonConfig?.marketExceptions?.formUpdateAccount || [
              "firstName",
              "lastName",
              "phoneNumber",
              "email"
          ];

    // Split fields into rows of 2
    const fieldRows: string[][] = [];
    for (let i = 0; i < formFields.length; i += 2) {
        fieldRows.push(formFields.slice(i, i + 2));
    }

    const renderField = (field: string): JSX.Element | null =>
        mapFieldToComponent({
            field,
            validationRules,
            translate,
            country,
            errors,
            control,
            disabled: true,
            type: "update"
        });
    return (
        <div css={styles}>
            <FormProvider {...form}>
                <form
                    className="form"
                    onSubmit={handleSubmit(onSubmit)}
                    noValidate
                >
                    {fieldRows.map(row => {
                        let rightCol: JSX.Element | null = null;
                        if (row[1]) {
                            rightCol = renderField(row[1]);
                        }
                        return (
                            <div
                                className="formRow"
                                key={`form-row-${row.join("-")}`}
                            >
                                <div className="formField halfWidth">
                                    {renderField(row[0])}
                                </div>
                                <div className="formField halfWidth">
                                    {rightCol}
                                </div>
                            </div>
                        );
                    })}

                    <button
                        type="submit"
                        className="continue-button"
                        data-testid="create_account_button"
                        disabled={isLoading}
                    >
                        {shouldUpdate
                            ? translate("step_update_your_account")
                            : translate("button_continue")}
                    </button>
                </form>
            </FormProvider>
        </div>
    );
}

export default UpdateAccountForm;
