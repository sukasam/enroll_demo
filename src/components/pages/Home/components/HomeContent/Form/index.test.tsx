import { fireEvent, render, screen } from "@testing-library/react";
import { Alpha2 } from "Constants/countryConfig/enums.js";
import { Customer } from "Hydra/getRefID";
import { useRouter } from "next/router";
import { act } from "react-dom/test-utils";
import Form from "../index";

jest.mock(
    "Components/shared/Form/components/TextField",
    () =>
        function MockTextField({
            name,
            label
        }: {
            name: string;
            label: string;
        }): JSX.Element {
            return <input data-testid={name} placeholder={label} />;
        }
);
jest.mock("Services/mixpanel/initializeMixPanel");
jest.mock(
    "Components/shared/PrimaryButton",
    () =>
        function MockPrimaryButton({
            children,
            ...props
        }: React.PropsWithChildren<any>): JSX.Element {
            return (
                <button {...props} type="button">
                    {children}
                </button>
            );
        }
);
jest.mock(
    "Components/shared/SelectDropdown",
    () =>
        function MockSelectDropdown({
            name,
            label,
            options
        }: {
            name: string;
            label: string;
            options: Array<{ label: string; value: string }>;
        }): JSX.Element {
            return (
                <select data-testid={name}>
                    <option value="">{label}</option>
                    {options.map(option => (
                        <option key={option.value} value={option.value}>
                            {option.label}
                        </option>
                    ))}
                </select>
            );
        }
);
jest.mock(
    "Components/shared/StyledCheckbox",
    () =>
        function MockStyledCheckbox({
            id,
            name,
            onChange
        }: {
            id: string;
            name: string;
            onChange: () => void;
        }): JSX.Element {
            return (
                <input
                    type="checkbox"
                    id={id}
                    name={name}
                    onChange={onChange}
                />
            );
        }
);
jest.mock(
    "Components/shared/Spinner",
    () =>
        function MockSpinner(): JSX.Element {
            return <div data-testid="spinner">Loading...</div>;
        }
);
jest.mock("Components/shared/Translate", () => ({
    useTranslate: (): ((key: string) => string) => (key: string) => key
}));
jest.mock("Contexts/UserContext", () => ({
    useUser: (): {
        setEnrollerId: jest.Mock;
        setEnrollerFullName: jest.Mock;
        setSponsorId: jest.Mock;
        setSponsorFullName: jest.Mock;
        enrollerId: string;
        sponsorFullName: string;
        enrollerFullName: string;
        isUserContextLoading: boolean;
    } => ({
        setEnrollerId: jest.fn(),
        setEnrollerFullName: jest.fn(),
        setSponsorId: jest.fn(),
        setSponsorFullName: jest.fn(),
        enrollerId: "",
        sponsorFullName: "",
        enrollerFullName: "",
        isUserContextLoading: false
    })
}));
jest.mock("Contexts/translation", () => ({
    useTranslations: (): {
        country: Alpha2;
        language: string;
        setCountry: jest.Mock;
        setLanguage: jest.Mock;
        fetchTranslations: jest.Mock;
    } => ({
        country: "US" as Alpha2,
        language: "en",
        setCountry: jest.fn(),
        setLanguage: jest.fn(),
        fetchTranslations: jest.fn()
    })
}));
jest.mock("next/router", () => ({
    useRouter: jest.fn()
}));
jest.mock("cookies-next", () => ({
    getCookie: jest.fn(),
    setCookie: jest.fn()
}));
jest.mock("./hooks", () => ({
    __esModule: true,
    default: (): {
        label: string;
        value: string;
    }[] => [{ label: "English", value: "en" }],
    checkRefID: jest.fn().mockResolvedValue("valid"),
    checkSponsor: jest.fn().mockResolvedValue("valid"),
    useDebounce: (value: any): any => value
}));

describe("Form", () => {
    beforeEach(() => {
        (useRouter as jest.Mock).mockImplementation(() => ({
            events: {
                on: jest.fn(),
                off: jest.fn()
            },
            query: {},
            push: jest.fn()
        }));
    });

    const defaultProps: { refId: string; referrer: Customer } = {
        refId: "",
        referrer: {} as Customer
    };

    it("renders correctly", () => {
        const { asFragment } = render(<Form {...defaultProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders country and language dropdowns", () => {
        render(<Form {...defaultProps} />);
        expect(screen.getByTestId("country")).toBeInTheDocument();
        expect(screen.getByTestId("language")).toBeInTheDocument();
    });

    it("renders enroller ID field", () => {
        render(<Form {...defaultProps} />);
        expect(screen.getByTestId("enrollerId")).toBeInTheDocument();
    });

    it("renders sponsor checkbox", () => {
        render(<Form {...defaultProps} />);
        expect(
            screen.getByLabelText("home_sponsor_different_enroller")
        ).toBeInTheDocument();
    });

    it("shows sponsor ID field when checkbox is checked", async () => {
        render(<Form {...defaultProps} />);
        const checkbox = screen.getByLabelText(
            "home_sponsor_different_enroller"
        );

        await act(async () => {
            fireEvent.click(checkbox);
        });

        expect(screen.getByTestId("sponsorId")).toBeInTheDocument();
    });
});
