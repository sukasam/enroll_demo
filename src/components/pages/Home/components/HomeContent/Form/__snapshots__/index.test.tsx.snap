// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<DocumentFragment>
  <div
    class="tile main css-1d9ev5j"
  >
    <div
      class="wrapper"
      data-testid="create_opportunity_section"
    >
      <div
        class="title"
        data-testid="home_create_your_opportunity"
      >
        home_create_your_opportunity
      </div>
      <div
        class="subtitle"
        data-testid="home_become_a_distributor"
      >
        home_become_a_distributor
      </div>
      <p
        data-testid="home_becoming_unicity_distributor"
      >
        home_becoming_unicity_distributor
      </p>
      <div
        class="css-tw5y6q"
      >
        <form>
          <select
            data-testid="country"
          >
            <option
              value=""
            >
              home_select_country
            </option>
            <option
              value="AU"
            >
              Australia
            </option>
            <option
              value="AT"
            >
              Austria - Österreich
            </option>
            <option
              value="BH"
            >
              Bahrain
            </option>
            <option
              value="BE"
            >
              Belgium - Belgique / België
            </option>
            <option
              value="BG"
            >
              Bulgaria - България
            </option>
            <option
              value="KH"
            >
              Cambodia
            </option>
            <option
              value="CA"
            >
              Canada
            </option>
            <option
              value="CO"
            >
              Colombia
            </option>
            <option
              value="HR"
            >
              Croatia - Hrvatska
            </option>
            <option
              value="CY"
            >
              Cyprus - Κύπρος
            </option>
            <option
              value="CZ"
            >
              Czech Republic - Česká republika
            </option>
            <option
              value="DK"
            >
              Denmark
            </option>
            <option
              value="DO"
            >
              Dominican Republic
            </option>
            <option
              value="EE"
            >
              Estonia - Eesti
            </option>
            <option
              value="FI"
            >
              Finland - Suomi
            </option>
            <option
              value="FR"
            >
              France
            </option>
            <option
              value="DE"
            >
              Germany - Deutschland
            </option>
            <option
              value="GR"
            >
              Greece - Ελλάδα
            </option>
            <option
              value="HK"
            >
              Hong Kong
            </option>
            <option
              value="HU"
            >
              Hungary - Magyarország
            </option>
            <option
              value="IS"
            >
              Iceland - Ísland
            </option>
            <option
              value="IN"
            >
              India
            </option>
            <option
              value="IE"
            >
              Ireland - Éire
            </option>
            <option
              value="IT"
            >
              Italy - Italia
            </option>
            <option
              value="JP"
            >
              Japan - 日本
            </option>
            <option
              value="JO"
            >
              Jordan
            </option>
            <option
              value="KW"
            >
              Kuwait
            </option>
            <option
              value="LA"
            >
              Laos
            </option>
            <option
              value="LV"
            >
              Latvia - Latvija
            </option>
            <option
              value="LT"
            >
              Lithuania - Lietuva
            </option>
            <option
              value="LU"
            >
              Luxembourg - Lëtzebuerg
            </option>
            <option
              value="MY"
            >
              Malaysia
            </option>
            <option
              value="MT"
            >
              Malta - Malta
            </option>
            <option
              value="MX"
            >
              Mexico
            </option>
            <option
              value="MA"
            >
              Morocco - Maroc
            </option>
            <option
              value="NL"
            >
              Netherlands - Nederland
            </option>
            <option
              value="NZ"
            >
              New Zealand
            </option>
            <option
              value="NO"
            >
              Norway - Norge
            </option>
            <option
              value="OM"
            >
              Oman
            </option>
            <option
              value="PH"
            >
              Philippines
            </option>
            <option
              value="PL"
            >
              Poland
            </option>
            <option
              value="PT"
            >
              Portugal - Portugal
            </option>
            <option
              value="PR"
            >
              Puerto Rico
            </option>
            <option
              value="QA"
            >
              Qatar
            </option>
            <option
              value="RO"
            >
              Romania - România
            </option>
            <option
              value="SA"
            >
              Saudi Arabia
            </option>
            <option
              value="SG"
            >
              Singapore
            </option>
            <option
              value="SK"
            >
              Slovakia - Slovensko
            </option>
            <option
              value="SI"
            >
              Slovenia - Slovenija
            </option>
            <option
              value="KR"
            >
              South Korea
            </option>
            <option
              value="ES"
            >
              Spain - España
            </option>
            <option
              value="SE"
            >
              Sweden - Sverige
            </option>
            <option
              value="CH"
            >
              Switzerland - Schweiz
            </option>
            <option
              value="TW"
            >
              Taiwan
            </option>
            <option
              value="TH"
            >
              Thailand
            </option>
            <option
              value="TR"
            >
              Turkey
            </option>
            <option
              value="AE"
            >
              United Arab Emirates
            </option>
            <option
              value="GB"
            >
              United Kingdom
            </option>
            <option
              value="US"
            >
              United States
            </option>
            <option
              value="VN"
            >
              Vietnam
            </option>
          </select>
          <select
            data-testid="language"
          >
            <option
              value=""
            >
              home_select_language
            </option>
            <option
              value="en"
            >
              English
            </option>
          </select>
          <div
            class="enroller-id-wrapper"
            data-testid="home_enroller_wrapper"
          >
            <input
              data-testid="enrollerId"
              placeholder="home_enroller"
            />
            <div
              class="enrollerName"
              data-testid="enroller_name"
            />
          </div>
          <div
            class="sponsor-checkbox"
          >
            <input
              id="useSponsor"
              name="useSponsor"
              type="checkbox"
            />
            <label
              data-testid="home_sponsor_different_enroller"
              for="useSponsor"
            >
              home_sponsor_different_enroller
            </label>
          </div>
          <button
            data-testid="home_get_started"
            disabled=""
            id="start_enrollment"
            type="button"
          >
            home_get_started
          </button>
        </form>
      </div>
      <div
        class="upgrade-text"
      >
        <a
          aria-label="Upgrade your account"
          class="upgrade-link"
          href="/login"
        >
          <div>
            home_already_have_account 
            
            <br />
             home_upgrade_to_distributor
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
